les/txrelay.go

```diff
@@ -68,93 +55,92 @@ func (self *lesTxRelay) Stop() {
 }
 
-func (self *lesTxRelay) unregisterPeer(p *peer) {
+// unregisterPeer implements peerSetNotify
+func (self *lesTxRelay) unregisterPeer(_ *peer) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
 
 	self.peerList = self.ps.AllPeers()
 }
 
-func (self *lesTxRelay) HasPeerWithEtherbase(etherbase *common.Address) error {
-	_, err := self.ps.getPeerWithEtherbase(etherbase)
-	return err
-}
-
-func (self *lesTxRelay) CanRelayTransaction(tx *types.Transaction) error {
-	// TODO(nategraf) self.gatewayFee is used in place of the minimum gateway fee among peers.
-	// When it is possible to query peers for their gateway fee, this should be replaced.
-
-	// Check if there we have peer accepting transactions without a gateway fee.
-	if self.gatewayFee.Cmp(common.Big0) <= 0 {
-		return nil
-	}
+func (self *lesTxRelay) CanRelayTransaction(tx *types.Transaction) bool {
+	self.lock.Lock()
+	defer self.lock.Unlock()
 
-	// Should have a peer that will accept and broadcast our transaction.
-	if err := self.HasPeerWithEtherbase(tx.GatewayFeeRecipient()); err != nil {
-		return err
-	}
-	if tx.GatewayFee().Cmp(self.gatewayFee) < 0 {
-		return errGatewayFeeTooLow
+	for _, p := range self.peerList {
+		if p.WillAcceptTransaction(tx) {
+			return true
+		}
 	}
-	return nil
+	return false
 }
 
 // send sends a list of transactions to at most a given number of peers at
 // once, never resending any particular transaction to the same peer twice
 func (self *lesTxRelay) send(txs types.Transactions) {
-	sendTo := make(map[*peer]types.Transactions)
-
 	for _, tx := range txs {
 		hash := tx.Hash()
-		_, ok := self.txSent[hash]
-		if !ok {
-			p, err := self.ps.getPeerWithEtherbase(tx.GatewayFeeRecipient())
-			// TODO(asa): When this happens, the nonce is still incremented, preventing future txs from being added.
-			// We rely on transactions to be rejected in light/txpool validateTx to prevent transactions
-			// with GatewayFeeRecipient != one of our peers from making it to the relayer.
-			if err != nil {
-				log.Error("Unable to find peer with matching etherbase", "err", err, "tx.hash", tx.Hash(), "tx.gatewayFeeRecipient", tx.GatewayFeeRecipient())
-				continue
-			}
-			sendTo[p] = append(sendTo[p], tx)
-			ltr := &ltrInfo{
-				tx:     tx,
-				sentTo: make(map[*peer]struct{}),
-			}
-			self.txSent[hash] = ltr
-			self.txPending[hash] = struct{}{}
+		if _, ok := self.txSent[hash]; ok {
+			continue
 		}
-	}
 
-	for p, list := range sendTo {
-		pp := p
-		ll := list
-		enc, _ := rlp.EncodeToBytes(ll)
+		self.txSent[hash] = tx
+		self.txPending[hash] = struct{}{}
 
+		// Send a single transaction per request to avoid failure coupling and
+		// because the expected base cost of a SendTxV2 request is 0, so it
+		// cost no extra to send multiple requests with one transaction each.
+		list := types.Transactions{tx}
+		enc, _ := rlp.EncodeToBytes(list)
+
+		// Assemble the request object with callbacks for the distributor.
 		reqID := genReqID()
 		rq := &distReq{
```

Comment:
iiuc, what's happening here is that we will relay the tx to every peer that accepts it? or the the first one that accepts it?
If it's the second case, that means we will wait for the response on every peer before trying with the second one..

What's the case here?